lib/msf/core/web_services/servlet/file_servlet.rb

```diff
@@ -0,0 +1,205 @@
+        result = { path: new_path }
+      else
+        result = { message: 'No such file' }
+      end
+      set_json_data_response(response: result)
+    }
+  end
+
+  # dir
+  def self.dir_entries
+    lambda {
+      warden.authenticate!
+      sanitized_params = sanitize_params(params, env['rack.request.query_hash'])
+      opts_path = sanitized_params[:path] || ''
+      path = File.join(Msf::Config.rest_files_directory, opts_path)
+      if safe_expand_path?(path) && File.directory?(path)
+        result = list_local_path(path)
+      else
+        result = { message: 'No such directory' }
+      end
+      set_json_data_response(response: result)
+    }
+  end
+
+  def self.dir_mkdir
+    lambda {
+      warden.authenticate!
+      opts = parse_json_request(request, true)
+      opts_path = opts[:path] || ''
+      path = File.join(Msf::Config.rest_files_directory, opts_path)
+      if safe_expand_path?(path)
+        FileUtils.mkdir_p(path)
+        result = { path: path }
+      else
+        result = { message: 'Failed to create folder' }
+      end
+      set_json_data_response(response: result)
+    }
+  end
+
+  def self.dir_delete
+    lambda {
+      warden.authenticate!
+      opts = parse_json_request(request, true)
+      opts_path = opts[:path] || ''
+      path = File.join(Msf::Config.rest_files_directory, opts_path)
+      if !rest_files_directory?(path) && safe_expand_path?(path) && File.directory?(path)
+        FileUtils.rm_rf(path)
+        result = { path: path }
+      else
+        result = { message: 'No such directory' }
+      end
+      set_json_data_response(response: result)
+    }
+  end
+
+  def self.dir_rename
+    lambda {
+      warden.authenticate!
+      opts = parse_json_request(request, true)
+      opts_path = opts[:path] || ''
+      opts_new_path = opts[:new_path] || ''
+      path = File.join(Msf::Config.rest_files_directory, opts_path)
+      new_path = File.join(Msf::Config.rest_files_directory, opts_new_path)
+      if (!rest_files_directory?(path) && safe_expand_path?(path) && File.directory?(path)) \
+        && (!rest_files_directory?(new_path) && safe_expand_path?(new_path))
+        FileUtils.mv(path, new_path)
+        result = { path: new_path }
+      else
+        result = { message: 'No such directory' }
+      end
+      set_json_data_response(response: result)
+    }
+  end
+end
+
+def list_local_path(path)
+  # Enumerate each item...
+  tbl = []
+  files = Dir.entries(path)
+  files.each do |file|
+    file_path = File.join(path, file)
+    stat = File.stat(file_path)
+    row = {
+      name: file,
+      type: stat.ftype || '',
+      size: stat.size ? stat.size.to_s : '',
+      last_modified: stat.mtime || ''
+    }
+    next unless file != '.' && file != '..'
+
+    tbl << row
+  end
+  return tbl
+end
+
+def safe_expand_path?(path)
+  current_directory = File.expand_path(Msf::Config.rest_files_directory)
+  tested_path = File.expand_path(path)
```

Comment:
`File.expand_path` does not append a path separator character to the end to ensure that we're limiting the operations to the expected path.

This should be:
```suggestion
  current_directory = File.expand_path(Msf::Config.rest_files_directory) + File::SEPARATOR
  tested_path = File.expand_path(path) + File::SEPARATOR
```